tests/test_control_low.py

```diff
@@ -207,6 +207,67 @@ def test_mqtt(self):
             oldmqtt["keep_alive_interval"],
         )
 
+    # Available from fw version 2.7.1
+    @vcr.use_cassette("tests/cassettes/TestControlInterface.test_color.yaml")
+    def test_color(self):
+        ctr = ControlInterface(self.host)
+
+        res = ctr.set_mode("color")._data
+        if self.isrecording:
+            print("set_mode:", res)
+        else:
+            self.assertEqual(res, {"code": 1000})
+
+        res = ctr.set_led_color_rgb(128, 128, 0)._data
+        if self.isrecording:
+            print("set_led_color_rgb:", res)
+        else:
+            self.assertEqual(res, {"code": 1000})
+
+        res = ctr.get_led_color()._data
+        if self.isrecording:
+            print("get_led_color:", res)
+        else:
+            self.assertEqual(
+                res,
+                {
+                    "hue": 60,
+                    "saturation": 255,
```

Comment:
For some reason on the device I tested this ([Model TWS250STP-BEU](https://xled-docs.readthedocs.io/en/latest/hardware.html#model-tws250stp-beu)) with FW version 2.8.3 this test fails - expects `255` but device returns `254`.

Reply:
@scrool, this is strange. I have the exact same model (250 leds, profile RGB), but with version 2.7.1. It is the one used to produce the cassettes. 
So what you say is that with firmware 2.8.3, when setting RGB to (128, 128, 0), the HSV becomes (60, 254, 128) ?
Whereas with version 2.7.1 the same input results in HSV (60, 255, 128). That's strange, since 255 should be the correct number: if any RGB component is zero (but not all), then saturation is at maximum, which is 255. Maybe the adjusted the calculation and accidentally introduced a rounding error?  
So I will upgrade my leds too and see if it changes here too. We know of course that different models and versions may differ in their output. But if this is a rounding error in the device, then I can try to find another test that is not "on the border" but gives the same value in both these versions.
(And then it suddenly  dawns on me, that RGBW and AWW models surely are different in these parameters. That is, the current documentation are limited to profile RGB right now, until someone with access to such devices can check what comes out.)

Reply:
That is strange indeed. I have downgraded FW on that device to 2.7.1 and that test still fails the same way.

Btw. I have also some RGBW here, I just didn't want to screw up my config :)

btw. using [shell wrappers](https://github.com/xled/xled-shell) it looks like this:

```
$ echo '{"red": 128, "green": 128, "blue": 0}' | ./post-json $IPADDR led/color && ./get $IPADDR led/color    
POST led/color
{
  "code": 1000
}
GET led/color
{
  "hue": 60,
  "saturation": 254,
  "value": 128,
  "red": 128,
  "green": 128,
  "blue": 0,
  "code": 1000
}
```

and to test that other way around:

```
$ echo '{"hue": 0, "saturation": 0, "value": 0}' | ./post-json $IPADDR led/color && ./get $IPADDR led/color
POST led/color
{
  "code": 1000
}
GET led/color
{
  "hue": 0,
  "saturation": 0,
  "value": 0,
  "red": 0,
  "green": 0,
  "blue": 0,
  "code": 1000
}
$ echo '{"hue": 60, "saturation": 255, "value": 128}' | ./post-json $IPADDR led/color && ./get $IPADDR led/color
POST led/color
{
  "code": 1000
}
GET led/color
{
  "hue": 60,
  "saturation": 255,
  "value": 128,
  "red": 128,
  "green": 128,
  "blue": 0,
  "code": 1000
}
```

Reply:
Yes, odd that it remains when you downgraded... it is contrary to what I get: 
I have now upgraded one of my two lights, both of exactly the same model TWS250STP-BEU. Then I tested on both, and the new firmware really has some problems with color. Check these identical sequences on my two devices:
```
>>> ctr1.firmware_version()._data
{'version': '2.7.1', 'code': 1000}
>>> ctr1.set_led_color_rgb(0, 255, 255)._data
{'code': 1000}
>>> ctr1.get_led_color()._data
{'hue': 180, 'saturation': 255, 'value': 255, 'red': 0, 'green': 255, 'blue': 255, 'code': 1000}
>>> ctr1.set_led_color_rgb(2, 255, 255)._data
{'code': 1000}
>>> ctr1.get_led_color()._data
{'hue': 180, 'saturation': 253, 'value': 255, 'red': 2, 'green': 255, 'blue': 255, 'code': 1000}
>>> ctr1.set_led_color_rgb(4, 255, 255)._data
{'code': 1000}
>>> ctr1.get_led_color()._data
{'hue': 180, 'saturation': 251, 'value': 255, 'red': 4, 'green': 255, 'blue': 255, 'code': 1000}
>>> 
>>> ctr2.firmware_version()._data
{'version': '2.8.3', 'code': 1000}
>>> ctr2.set_led_color_rgb(0, 255, 255)._data
{'code': 1000}
>>> ctr2.get_led_color()._data
{'hue': 180, 'saturation': 254, 'value': 255, 'red': 1, 'green': 255, 'blue': 255, 'code': 1000}
>>> ctr2.set_led_color_rgb(2, 255, 255)._data
{'code': 1000}
>>> ctr2.get_led_color()._data
{'hue': 180, 'saturation': 252, 'value': 255, 'red': 3, 'green': 255, 'blue': 255, 'code': 1000}
>>> ctr2.set_led_color_rgb(4, 255, 255)._data
{'code': 1000}
>>> ctr2.get_led_color()._data
{'hue': 180, 'saturation': 250, 'value': 255, 'red': 5, 'green': 255, 'blue': 255, 'code': 1000}
```
The previous firmware is consistent and gives the expected results. The new firmware *adds one* to the smallest rgb coordinate ("red" in this case) compared to what I asked for, and correspondingly *decreases* the expected saturation by one. This seems to happen when I set red and green to 255. When I use a lower number, say 250, it behaves better.

What I have done now: I found two settings, one of rgb and one of hsv, which actually behaves the same on both devices and gives the intuitively correct result on both. I have modified the unit test to use those values, and have run it on both my devices, without cassettes (but without isrecording), and it passes both here. I will check it in.

(How to carefully design a unit test not to trip a bug... Well, it is not our task to detect inconsistencies in the firmware, only in our code.)